File: src/checks/checkout.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

export const checkoutCreateCheck: Check = {
  name: 'Checkout Create',
  description: 'Create checkout session',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.testCustomerId || ctx.createdCustomerId;

    if (!customerId) {
      return {
        name: 'Checkout Create',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      // Access the SDK directly to call checkout
      const sdk = (client as unknown as { sdk: { checkout?: (data: { customerId?: string; amount: number; returnUrl: string }) => Promise<{ id: string; url: string }> } }).sdk;

      if (!sdk.checkout) {
        // Endpoint may not exist - skip gracefully
        const duration = performance.now() - start;
        return {
          name: 'Checkout Create',
          success: true,
          duration,
          message: 'Skipped (checkout endpoint not available)',
          details: 'The checkout method is not available in the SDK',
        };
      }

      const result = await sdk.checkout({
        customerId,
        amount: 1000, // $10.00 in cents
        returnUrl: 'https://example.com/checkout/success',
      });
      const duration = performance.now() - start;

      return {
        name: 'Checkout Create',
        success: true,
        duration,
        message: `Session: ${result.id}`,
        details: `URL: ${result.url}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      // Handle 404/501 gracefully for endpoints that may not exist
      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Checkout Create',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Checkout Create',
        success: false,
        duration,
        message: err.message || 'Failed to create checkout session',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/checkout.py
"""Checkout operation checks."""
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _checkout_create_check(ctx: CheckContext) -> CheckResult:
    """Create checkout session."""
    customer_id = ctx.created_customer_id or ctx.test_customer_id
    if not customer_id:
        return CheckResult(
            name="checkout_create",
            success=False,
            duration=0,
            message="No customer ID available",
            suggestion="Run customer_create check first or set TEST_CUSTOMER_ID"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'checkout') and not hasattr(client, 'create_checkout'):
            return CheckResult(
                name="checkout_create",
                success=True,
                duration=0,
                message="Skipped (checkout not available)",
                details="The checkout method is not available in the SDK"
            )

        checkout_method = getattr(client, 'checkout', None) or getattr(client, 'create_checkout', None)

        result = checkout_method(
            customer_id=customer_id,
            amount=1000,  # $10.00 in cents
            return_url="https://example.com/checkout/success"
        )

        session_id = getattr(result, 'id', str(result))
        url = getattr(result, 'url', 'N/A')

        return CheckResult(
            name="checkout_create",
            success=True,
            duration=0,
            message=f"Session: {session_id}",
            details=f"URL: {url[:50]}..." if len(str(url)) > 50 else f"URL: {url}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="checkout_create",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="checkout_create",
            success=False,
            duration=0,
            message=f"Failed to create checkout: {e}"
        )


checkout_create_check = Check(
    name="checkout_create",
    description="Create checkout session",
    run=_checkout_create_check
)
